File: solutions/csharp/grade-school/1/GradeSchool.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class GradeSchool
{
    private SortedDictionary<int, List<string>> _roster = new SortedDictionary<int, List<string>>();
    
    public bool Add(string student, int grade)
    {
        if (Roster().Contains(student))
        {
            return false;
        }
        else
        {
            if (!_roster.ContainsKey(grade))
            {
                _roster[grade] = new List<string>();
            }
            _roster[grade].Add(student);
            _roster[grade].Sort();
            return true;
        }
    }

    public IEnumerable<string> Roster()
    {
        
        return _roster.Values.SelectMany(grade => grade);
    }

    public IEnumerable<string> Grade(int grade)
    {
        if (!_roster.ContainsKey(grade))
            return Array.Empty<string>();
        return _roster[grade];
    }
}

File: solutions/fsharp/grade-school/1/GradeSchool.fs
module GradeSchool

type School = Map<int, string list>

let empty: School = Map.empty

let add (student: string) (grade: int) (school: School): School = 
    if Map.exists (fun g s -> List.contains student s) school then
        school
    else
        match (Map.tryFind grade school) with
        | None -> Map.add grade [student] school
        | Some _ -> Map.add grade (List.sort (student::school.[grade])) school

let roster (school: School): string list = Map.values school |> List.concat |> List.distinct

let grade (number: int) (school: School): string list = 
    match (Map.tryFind number school) with
    | None -> []
    | Some _ -> school.[number]
